### PR TITLE
Add option to pad after applying ufunc, and reroute cumsum

### LINKED ISSUE #458: [FEATURE] Rewiring `cumsum` via (parts of) the grid_ufunc logic.
While working on https://github.com/jbusecke/xgcm/tree/vector_padding I am encountering some test failures, specifically the `xgcm/test/test_grid.py::test_keep_coords[xx-cumint]`, `xgcm/test/test_grid.py::test_keep_coords[xx-cumsum]` tests which are testing the consistency of output coordinates. 

While checking on this I noticed that the `Grid.cumsum` method in the `grid_ufunc_refactor` branch still calls the `Axis.cumsum` method, which has a bunch of [custom logic for padding/rewrapping the output](https://github.com/jbusecke/xgcm/blob/c742c47dbfee0a48b034baec965d33ef3fb21579/xgcm/grid.py#L702-L721) ([which itself is simply xarray.cumsum()](https://github.com/jbusecke/xgcm/blob/c742c47dbfee0a48b034baec965d33ef3fb21579/xgcm/grid.py#L700)). 
We will ultimately have to refactor this logic into the Grid method, when we remove the Axis methods. 

Should/Can this logic be implemented via the gufunc logic? That would solve my issue here, since [these changes I made](https://github.com/jbusecke/xgcm/blob/c742c47dbfee0a48b034baec965d33ef3fb21579/xgcm/grid_ufunc.py#L994-L1026) would ensure that the treatment of output coordinates is always consistent.

### PR DESCRIPTION
**Pull request by TomNicholas**
Cumsum is an annoying special case (see #458), because it has to be padded after the ufunc is applied. This PR so far generalises `apply_as_grid_ufunc` with a new `pad_before_func` kwarg that defaults to True, but if False will instead pad after applying the ufunc. 

To add that required factorising a lot of the code in `apply_as_grid_ufunc`, which is fine except that now I have all these internal functions with huge argument lists. I'm thinking about using [some refactoring approach to remove all of those](https://refactoring.guru/smells/long-parameter-list).

 - [x] Closes #458
 - [x] Tests added
 - [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by TomNicholas** [comment]
@jbusecke I've encounter a problem where doing first `apply ufunc` then `pad` fails because the apply_ufunc strips some dimension coordinates which the padding function then expects to be present in order to pad.

(This isn't a problem the other way around because padding first will pad the input dataarray with all its coordinates still present.)

Q: Will your padding PR help with this in any way?

**Comment by jbusecke** [comment]
It potentially could, but my suspicion is that it would make things a lot more complicated. I do not understand why coordinates are needed to pad though? It should simply add a row of values to one side and reindex?
Or am I missing something?
Can we use the np.pad function instead or modify the behavior of xr.pad for dataarrays without dimension coordinates?

**Comment by jbusecke** [comment]
I just had a few tests failing over at #459 and went through the code in `_pad_array`. I noticed this restriction here: https://github.com/xgcm/xgcm/blob/83deab5c6abbca2de408502e4e26760a618b7d98/xgcm/duck_array_ops.py#L132-L133

I interpret this as 'periodic boundary padding is not allowed for cumsum'? Is that right? If that is true, we would need to build in some checks to make sure that that case is never run. Just wanted to pin this. Maybe @rabernat has some more insight?

**Comment by jbusecke** [comment]
Reminder to take care of [these cumsum tests](https://github.com/xgcm/xgcm/pull/459#issuecomment-1093065205) and make sure they pass. Happy to help with that too.

**Comment by TomNicholas** [comment]
> problem where doing first apply ufunc then pad fails because the apply_ufunc strips some dimension coordinates which the padding function then expects to be present in order to pad.

This seems to have gone away now, and the tests are passing locally :champagne: 

There might be some ways to refactor the code inside `grid_ufunc.py` to be neater though.

**Comment by TomNicholas** [comment]
@jbusecke if you're happy with this then we can merge it

File: xgcm/gridops.py [hunk 1]
```diff
@@ -205,3 +214,67 @@ def max_center_to_inner(a):
...
+
+
+@as_grid_ufunc(
+    signature="(X:center)->(X:left)",
+    boundary_width={"X": (1, 0)},
+    fill_value=0,
+    pad_before_func=False,
+)
+def cumsum_center_to_left(a):
+    # TODO will this dispatch to dask.array.cumsum if it encounters a dask array?
```
Comment (jbusecke) [inline xgcm/gridops.py:229]:
```suggestion
```

**Review by jbusecke — APPROVED** [review]
This looks all good in my eyes. Lets merge

File: xgcm/grid.py [hunk 2]
```diff
@@ -2286,7 +2288,7 @@ def cumsum(self, da, axis, **kwargs):
 
         >>> grid.max(da, ["X", "Y"], fill_value={"X": 0, "Y": 100})
         """
-        return self._grid_func("cumsum", da, axis, **kwargs)
+        return self._1d_grid_ufunc_dispatch("cumsum", da, axis, **kwargs)
```
[the remark this answers is not in the document]
Comment (TomNicholas) [reply xgcm/grid.py:2227]: Yes we can, just removed it!

[the remark this answers is not in the document]
Comment (jbusecke) [reply xgcm/grid.py:2227]: Amazing!

**Pull request merged**